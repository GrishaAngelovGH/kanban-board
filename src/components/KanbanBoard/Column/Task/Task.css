.task-description {
  max-height: 200px;
}

.move-btn:hover {
  color: #fff;
  background-color: #808080;
}
