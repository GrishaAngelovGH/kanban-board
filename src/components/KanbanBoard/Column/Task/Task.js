import { useDraggable } from "@dnd-kit/core"

import OverlayTrigger from "react-bootstrap/OverlayTrigger"
import Tooltip from "react-bootstrap/Tooltip"

import users from "users"

const Task = ({ id, columnId, assignedIds, title, description, onEdit, onAssignUser, onDelete }) => {
  const { attributes, listeners, setNodeRef, transform } = useDraggable({
    id,
    data: { description, columnId }
  })

  const style = transform ? {
    transform: `translate3d(${transform.x}px, ${transform.y}px, 0)`,
  } : undefined

  const handleDelete = () => {
    onDelete(id, columnId)
  }

  const handleEdit = () => {
    onEdit({ id, title, description }, columnId)
  }

  const handleAssignUser = () => {
    onAssignUser({ id, title, description, assignedIds }, columnId)
  }

  return (
    <div ref={setNodeRef} style={style} className="row bg-white mt-3 rounded shadow p-1">
      <div className="col-9">
        <p className="fw-bold">{title}</p>
      </div>
      <div className="col-3">
        <i {...listeners} {...attributes} className="bi bi-grip-horizontal fs-2"></i>
      </div>
      <p className="text-secondary">{description}</p>

      <div className="mb-3">
        {
          assignedIds.map(v => {
            const user = users.find(u => u.id === v)

            const renderTooltip = (props) => (
              <Tooltip {...props}>
                {user.name}
              </Tooltip>
            )

            return (
              <OverlayTrigger
                key={v}
                placement="top"
                delay={{ show: 250, hide: 400 }}
                overlay={renderTooltip}
              >
                <img src={user.image} width={35} className="rounded-circle me-1" alt="assigned-user" />

              </OverlayTrigger>
            )
          })
        }
      </div>

      <div className="d-flex">
        <i role="button" onClick={handleDelete} className="bi bi-trash fs-4 text-danger"></i>
        <i role="button" onClick={handleEdit} className="bi bi-pencil-square fs-4 text-secondary mx-1"></i>
        <i role="button" onClick={handleAssignUser} className="bi bi-person-circle fs-4 text-secondary mx-1"></i>
      </div>
    </div>
  )
}

export default Task