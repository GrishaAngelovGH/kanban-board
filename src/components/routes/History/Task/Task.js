import RichTextDescription from "components/RichTextDescription"
import Button from "react-bootstrap/Button"

import history from "persistent/history"

const Task = ({ id, title, description, priority, column, showToastWithMessage }) => {
  const handleRestore = () => {
    history.restoreTask(id, column.id)
    showToastWithMessage("The task is successfully restored")
  }

  const handleDelete = () => {
    history.removeTask(id, column.id)
    showToastWithMessage("The task is successfully removed")
  }

  return (
    <div className="mt-3 border border-3 shadow rounded p-3 bg-light">
      <h3>{title}</h3>

      <RichTextDescription description={description} className="mb-3" />

      {
        priority.length > 0 && (
          <p className="text-capitalize">Priority: {priority}</p>
        )
      }

      <div className="task-actions">
        <Button
          variant="primary"
          size="sm"
          onClick={handleRestore}
        >
          Restore
        </Button>
        <Button
          variant="danger"
          size="sm"
          onClick={handleDelete}
        >
          Delete
        </Button>
      </div>
    </div>
  )
}

export default Task