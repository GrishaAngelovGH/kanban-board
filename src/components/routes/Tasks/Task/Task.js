import RichTextDescription from "components/RichTextDescription"
import Collaborators from "components/routes/Tasks/Collaborators"
import MoveToColumnDropdownButton from "components/routes/Tasks/MoveToColumnDropdownButton"
import Button from "react-bootstrap/Button"

import boardRepository from "persistent/persistentKanbanBoardRepository"

const Task = ({
  id: taskId, userId, title, description, priority, assignedIds,
  isLocked, column, setActiveKey, showToastWithMessage
}) => {
  const handleUnassignUser = () => {
    boardRepository.removeAssignedUserFromTask(taskId, column.id, userId)
    setActiveKey("")
    showToastWithMessage(`The user is successfully unassigned from "${title}" task`)
  }

  const handleUnassignCollaborators = () => {
    assignedIds.forEach(id => {
      id !== userId && boardRepository.removeAssignedUserFromTask(taskId, column.id, id)
    })

    setActiveKey("")
    showToastWithMessage(`The remaining collaborators have been successfully unassigned from task "${title}"`)
  }

  return (
    <div className="mt-3 border border-3 shadow rounded p-3 bg-light">
      <h3>{title} {isLocked && <i className="bi bi-lock"></i>}</h3>

      <RichTextDescription description={description} className="mb-3" />

      {
        priority.length > 0 && (
          <p className="text-capitalize">Priority: {priority}</p>
        )
      }

      <Collaborators ids={assignedIds} userId={userId} />

      <div className="task-actions">
        <MoveToColumnDropdownButton
          column={column}
          task={{ taskId, isLocked }}
          showToastWithMessage={showToastWithMessage}
          setActiveKey={setActiveKey}
        />

        <Button
          variant="danger"
          size="sm"
          onClick={handleUnassignUser}
        >
          Unassign Me
        </Button>

        <Button
          variant="danger"
          size="sm"
          onClick={handleUnassignCollaborators}
        >
          Leave only me
        </Button>
      </div>
    </div>
  )
}

export default Task