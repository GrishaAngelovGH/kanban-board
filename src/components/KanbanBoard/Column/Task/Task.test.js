import { render } from "@testing-library/react"

import Task from "./Task"

test("should render Task component", () => {
  const view = render(<Task title="Task Title" description="Task Description" />)

  expect(view).toMatchSnapshot()
})