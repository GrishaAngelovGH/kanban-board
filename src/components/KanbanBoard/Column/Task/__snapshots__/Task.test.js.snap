// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Task component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row bg-white mt-3 rounded shadow p-1"
        style="transform: translate3d(0px, 0px, 0);"
      >
        <div
          class="col-9"
        >
          <p
            class="fw-bold"
          >
            Task Title
          </p>
        </div>
        <div
          class="col-3"
        >
          <i
            aria-describedby=""
            aria-disabled="false"
            aria-pressed="true"
            aria-roledescription="draggable"
            class="bi bi-grip-horizontal fs-2"
            role="button"
            tabindex="0"
          />
        </div>
        <p
          class="text-secondary"
        >
          Task Description
        </p>
        <div
          class="mb-3"
        />
        <div
          class="d-flex"
        >
          <i
            class="bi bi-trash fs-4 text-danger"
            role="button"
          />
          <i
            class="bi bi-pencil-square fs-4 text-secondary mx-1"
            role="button"
          />
          <i
            class="bi bi-person-circle fs-4 text-secondary mx-1"
            role="button"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row bg-white mt-3 rounded shadow p-1"
      style="transform: translate3d(0px, 0px, 0);"
    >
      <div
        class="col-9"
      >
        <p
          class="fw-bold"
        >
          Task Title
        </p>
      </div>
      <div
        class="col-3"
      >
        <i
          aria-describedby=""
          aria-disabled="false"
          aria-pressed="true"
          aria-roledescription="draggable"
          class="bi bi-grip-horizontal fs-2"
          role="button"
          tabindex="0"
        />
      </div>
      <p
        class="text-secondary"
      >
        Task Description
      </p>
      <div
        class="mb-3"
      />
      <div
        class="d-flex"
      >
        <i
          class="bi bi-trash fs-4 text-danger"
          role="button"
        />
        <i
          class="bi bi-pencil-square fs-4 text-secondary mx-1"
          role="button"
        />
        <i
          class="bi bi-person-circle fs-4 text-secondary mx-1"
          role="button"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
